README.md
# Zadanie sj - LittleXML

Pre spustenie projektu treba pustit prikaz: 

```sh
cabal run
```

___treba mat pravdaze nainstalovany Haskell a Cabal a potrebne dependencie alebo otvorit projekt v devcontaineri.___

Vstupny subor je input.txt 

## Uloha 1 - priklady jazyka

Vytvorili sme tieto 3 priklady LittleXML validnych dokumentov: 

```xml
<?xml version=2.3 ?>
<simple_element/>
```

```xml
<?xml version=2.3 ?>
<hoco>
Ahoj ako mas ?
</hoco>
```

```xml
<hoco>
<inner_element>
<simple_element/>
</inner_element>
</hoco>
```

## Uloha 2

Terminaly sme nahradili tymito tokenmi: 

```
" <?xml" = XML_OPEN_TK
"version =" = XML_VERSION_TK
"?>" = XML_CLOSE_TK
"." = DOT_TK
" <" = SIMPLE_START_TK
" >" = SIMPLE_CLOSE_TK
" </" = SLASHED_START_TK
"/>" = SLASHED_CLOSE_TK
"_" = UNDERSCORE_TK
" :" = COLON_TK
"-" = DASH_TK
A-Za-z = LETTER_TK
0-9 = DIGIT_TK
```

Vysledna gramatika pred prevedenim do LL1 je takato: 

```
xmldokument -> element | xmldecl element
xmldecl -> XML_OPEN_TK XML_VERSION_TK vernumb XML_CLOSE_TK
vernumb -> number DOT_TK number
element -> emptyelemtag | starttag words endtag | starttag elements endtag    <-- lavy faktor: starttag
starttag -> SIMPLE_START_TK name SIMPLE_CLOSE_TK
endtag -> SLASHED_START_TK name SIMPLE_CLOSE_TK
words -> word words | 𝝴
elements -> element | 𝝴
emptyelemtag -> SIMPLE_START_TK name SLASHED_CLOSE_TK
name -> letter | UNDERSCORE_TK | COLON_TK | letter namechars |  UNDERSCORE_TK namechars | COLON_TK namechars  <-- lavy faktor: letter, UNDERSCORE_TK, COLON_TK
namechars -> namechar namechars | 𝝴
namechar -> letter | digit | DOT_TK | UNDERSCORE_TK | DASH_TK | COLON_TK
letter -> LETTER_TK
number -> digit | digit number <-- lavy faktor: digit
digit -> DIGIT_TK
word -> char | char word <-- lavy faktor char
char -> letter | digit | UNDERSCORE_TK | COLON_TK | DASH_TK ... (okrem <>)
```

## Uloha 3 - transformacia na LL1

1. krok lavej faktorizacie

```
xmldokument -> element | xmldecl element
xmldecl -> XML_OPEN_TK XML_VERSION_TK vernumb XML_CLOSE_TK
vernumb -> number DOT_TK number
element -> emptyelemtag | starttag elementshit
elementshit -> words endtag | elements endtag
starttag -> SIMPLE_START_TK name SIMPLE_CLOSE_TK
endtag -> SLASHED_START_TK name SIMPLE_CLOSE_TK
words -> word words | 𝝴
elements -> element | 𝝴
emptyelemtag -> SIMPLE_START_TK name SLASHED_CLOSE_TK
name -> letter | UNDERSCORE_TK | COLON_TK | letter namechars |  UNDERSCORE_TK namechars | COLON_TK namechars  <-- lavy faktor: letter, UNDERSCORE_TK, COLON_TK
namechars -> namechar namechars | 𝝴
namechar -> letter | digit | DOT_TK | UNDERSCORE_TK | DASH_TK | COLON_TK
letter -> LETTER_TK
number -> digit | digit number <-- lavy faktor: digit
digit -> DIGIT_TK
word -> char | char word <-- lavy faktor char
char -> letter | digit | UNDERSCORE_TK | COLON_TK | DASH_TK ... (okrem <>)
```

2. krok lavej faktorizacie

```
xmldokument -> element | xmldecl element
xmldecl -> XML_OPEN_TK XML_VERSION_TK vernumb XML_CLOSE_TK
vernumb -> number DOT_TK number
element -> emptyelemtag | starttag elementshit
elementshit -> words endtag | elements endtag
starttag -> SIMPLE_START_TK name SIMPLE_CLOSE_TK
endtag -> SLASHED_START_TK name SIMPLE_CLOSE_TK
words -> word words | 𝝴
elements -> element | 𝝴
emptyelemtag -> SIMPLE_START_TK name SLASHED_CLOSE_TK
name -> letter nameshit | UNDERSCORE_TK nameshit | COLON_TK nameshit
nameshit -> namechars | 𝝴
namechars -> namechar namechars | 𝝴
namechar -> letter | digit | DOT_TK | UNDERSCORE_TK | DASH_TK | COLON_TK
letter -> LETTER_TK
number -> digit | digit number <-- lavy faktor: digit
digit -> DIGIT_TK
word -> char | char word <-- lavy faktor char
char -> letter | digit | UNDERSCORE_TK | COLON_TK | DASH_TK ... (okrem <>)
```

3. krok lavej faktorizacie

```
xmldokument -> element | xmldecl element
xmldecl -> XML_OPEN_TK XML_VERSION_TK vernumb XML_CLOSE_TK
vernumb -> number DOT_TK number
element -> emptyelemtag | starttag elementshit
elementshit -> words endtag | elements endtag
starttag -> SIMPLE_START_TK name SIMPLE_CLOSE_TK
endtag -> SLASHED_START_TK name SIMPLE_CLOSE_TK
words -> word words | 𝝴
elements -> element | 𝝴
emptyelemtag -> SIMPLE_START_TK name SLASHED_CLOSE_TK
name -> letter nameshit | UNDERSCORE_TK nameshit | COLON_TK nameshit
nameshit -> namechars | 𝝴
namechars -> namechar namechars | 𝝴
namechar -> letter | digit | DOT_TK | UNDERSCORE_TK | DASH_TK | COLON_TK
letter -> LETTER_TK
number -> digit numbershit
numbershit -> number | 𝝴
digit -> DIGIT_TK
word -> char | char word <-- lavy faktor char
char -> letter | digit | UNDERSCORE_TK | COLON_TK | DASH_TK ... (okrem <>)
```

4. krok lavej faktorizacie

```
xmldokument -> element | xmldecl element
xmldecl -> XML_OPEN_TK XML_VERSION_TK vernumb XML_CLOSE_TK
vernumb -> number DOT_TK number
element -> emptyelemtag | starttag elementshit
elementshit -> words endtag | elements endtag
starttag -> SIMPLE_START_TK name SIMPLE_CLOSE_TK
endtag -> SLASHED_START_TK name SIMPLE_CLOSE_TK
words -> word words | 𝝴
elements -> element | 𝝴
emptyelemtag -> SIMPLE_START_TK name SLASHED_CLOSE_TK
name -> letter nameshit | UNDERSCORE_TK nameshit | COLON_TK nameshit
nameshit -> namechars | 𝝴
namechars -> namechar namechars | 𝝴
namechar -> letter | digit | DOT_TK | UNDERSCORE_TK | DASH_TK | COLON_TK
letter -> LETTER_TK
number -> digit numbershit
numbershit -> number | 𝝴
digit -> DIGIT_TK
word -> char  wordshit
wordshit -> word | 𝝴
char -> letter | digit | UNDERSCORE_TK | COLON_TK | DASH_TK ... (okrem <>)
```


5. Vyhodenie zbytocneho pravidla nameshit

```
xmldokument -> element | xmldecl element
xmldecl -> XML_OPEN_TK XML_VERSION_TK vernumb XML_CLOSE_TK
vernumb -> number DOT_TK number
element -> emptyelemtag | starttag elementshit
elementshit -> words endtag | elements endtag
starttag -> SIMPLE_START_TK name SIMPLE_CLOSE_TK
endtag -> SLASHED_START_TK name SIMPLE_CLOSE_TK
words -> word words | 𝝴
elements -> element | 𝝴
emptyelemtag -> SIMPLE_START_TK name SLASHED_CLOSE_TK
name -> letter namechars | UNDERSCORE_TK namechars | COLON_TK namechars
namechars -> namechar namechars | 𝝴
namechar -> letter | digit | DOT_TK | UNDERSCORE_TK | DASH_TK | COLON_TK
letter -> LETTER_TK
number -> digit numbershit
numbershit -> number | 𝝴
digit -> DIGIT_TK
word -> char  wordshit
wordshit -> word | 𝝴
char -> letter | digit | UNDERSCORE_TK | COLON_TK | DASH_TK ... (okrem <>)
```

6. dalsie upravy
 - vyhodili sme neterminaly, ktore idu do jedneho terminalu digit, letter...)
 - vytiahli sme SIMPLE_START_TK pomocou elementshitshit
 - vyhodili sme uplne words neterminal

### Vysledna LL1 gramatika

```
xmldokument -> element
            | xmldecl element
xmldecl -> XML_OPEN_TK XML_VERSION_TK vernumb XML_CLOSE_TK
vernumb -> number DOT_TK number
element -> SIMPLE_START_TK name elementshitshit
elementshitshit ->  SLASHED_CLOSE_TK
            | SIMPLE_CLOSE_TK elementshit
elementshit -> word endtag
            | elements endtag
endtag -> SLASHED_START_TK name SIMPLE_CLOSE_TK
elements -> element
            | 𝝴
name -> LETTER_TK namechars
            | UNDERSCORE_TK namechars
            | COLON_TK namechars
namechars -> namechar namechars
            | 𝝴
namechar -> LETTER_TK
            | DIGIT_TK
            | DOT_TK
            | UNDERSCORE_TK
            | DASH_TK
            | COLON_TK
number -> DIGIT_TK numbershit
numbershit -> number
            | 𝝴
word -> char wordshit
wordshit -> word
            | 𝝴
char -> LETTER_TK
            | DIGIT_TK
            | UNDERSCORE_TK
            | COLON_TK
            | DASH_TK
```

## Uloha c.4

### FIRST1

```
F1(xmldokument) = F1(element) -  𝝴  U F1(xmldecl) - 𝝴 = { XML_OPEN_TK, SIMPLE_START_TK }
F1(xmldecl) =  { XML_OPEN_TK }
F1(vernumb) = F1(number) - 𝝴  = { DIGIT_TK }
F1(element) = { SIMPLE_START_TK }
F1(elementshitshit) = { SLASHED_CLOSE_TK, SIMPLE_CLOSE_TK }
F1(elementshit) = F1(word) - 𝝴  U F1(elements) - 𝝴 U F1(endtag) = { SLASHED_START_TK, SIMPLE_START_TK, LETTER_TK, DIGIT_TK, DOT_TK, UNDERSCORE_TK, DASH_TK, COLON_TK ... okrem <> }
F1(endtag) = { SLASHED_START_TK }
F1(elements) = F1(element) - 𝝴 U { 𝝴 } = { SIMPLE_START_TK, 𝝴 }
F1(name) = { LETTER_TK, UNDERSCORE_TK, COLON_TK }
F1(namechars) = F1(namechar) - 𝝴 U { 𝝴 } = { 𝝴, LETTER_TK, DIGIT_TK, DOT_TK, UNDERSCORE_TK, DASH_TK, COLON_TK ... okrem <> }
F1(namechar) = { LETTER_TK, DIGIT_TK, DOT_TK, UNDERSCORE_TK, DASH_TK, COLON_TK  ... okrem <>  }
F1(number) = { DIGIT_TK }
F1(numbershit) = F1(number) - 𝝴 U { 𝝴 }  = { DIGIT_TK, 𝝴 }
F1(word) = F1(char) - 𝝴 = { LETTER_TK, DIGIT_TK, DOT_TK, UNDERSCORE_TK, DASH_TK, COLON_TK ... okrem <> }
F1(wordshit) = F1(word) U { 𝝴 } = { 𝝴, LETTER_TK, DIGIT_TK, DOT_TK, UNDERSCORE_TK, DASH_TK, COLON_TK ... okrem <> }
F1(char) = { LETTER_TK, DIGIT_TK, DOT_TK, UNDERSCORE_TK, DASH_TK, COLON_TK ... okrem <> }
```

### FOLLOW1

```
FO1(xmldokument) = { $ }
FO1(xmldecl) = F1(element) - 𝝴 = { SIMPLE_START_TK }
FO1(vernum) = { XML_CLOSE_TK }
FO1(element) = FO1(xmldokument) U FO1(elements) = { $, SLASHED_START_TK }
FO1(elementshitshit) = FO1(element) = { $, SLASHED_START_TK }
FO1(elementshit) = FO1(elementshitshit) = { $, SLASHED_START_TK }
FO1(endtag) = FO1(elementshit) = { $, SLASHED_START_TK }
FO1(elements) = F1(endtag) - 𝝴 = { SLASHED_START_TK }
FO1(name) = { SIMPLE_CLOSE_TK } U F1(elementshitshit) - 𝝴 = { SLASHED_CLOSE_TK, SIMPLE_CLOSE_TK }
FO1(namechars) = FO1(name) = { SLASHED_CLOSE_TK, SIMPLE_CLOSE_TK }
FO1(namechar) = F1(namechars) - 𝝴 U FO1(namechars) = { SLASHED_CLOSE_TK, SIMPLE_CLOSE_TK, LETTER_TK, DIGIT_TK, DOT_TK, UNDERSCORE_TK, DASH_TK, COLON_TK ... okrem <> }
FO1(number) = FO1(numbershit) U { DOT_TK } U FO1(vernumb) = { DOT_TK, XML_CLOSE_TK }
FO1(numbershit) = FO1(number) = { DOT_TK, XML_CLOSE_TK }
FO1(word) = FO1(wordshit) U F1(endtag) - 𝝴 = { SLASHED_START_TK }
FO1(wordshit) = FO1(word) = { SLASHED_START_TK }
FO1(char) = F1(wordshit) - 𝝴 U FO1(word) =  { SLASHED_START_TK, LETTER_TK, DIGIT_TK, DOT_TK, UNDERSCORE_TK, DASH_TK, COLON_TK ... okrem <> }
```

Je to LL1! 

## 5. Tabulka prechodov

Ocislovane pravidla: 

```
1. xmldokument -> element
2. xmldokument -> xmldecl element
3. xmldecl -> XML_OPEN_TK XML_VERSION_TK vernumb XML_CLOSE_TK
4. vernumb -> number DOT_TK number
5. element -> SIMPLE_START_TK name elementshitshit
6. elementshitshit ->  SLASHED_CLOSE_TK
7. elementshitshit ->  SIMPLE_CLOSE_TK elementshit
8. elementshit -> word endtag
9. elementshit -> elements endtag
10. endtag -> SLASHED_START_TK name SIMPLE_CLOSE_TK
11. elements -> element
12. elements -> 𝝴
13. name -> LETTER_TK namechars
14. name -> UNDERSCORE_TK namechars
15. name -> COLON_TK namechars
16. namechars -> namechar namechars
17. namechars -> 𝝴
18. namechar -> LETTER_TK
19. namechar -> DIGIT_TK
20. namechar -> DOT_TK
21. namechar -> UNDERSCORE_TK
22. namechar -> DASH_TK
23. namechar -> COLON_TK
24. number -> DIGIT_TK numbershit
25. numbershit -> number
26. numbershit -> 𝝴
27. word -> char wordshit
28. wordshit -> word
29. wordshit -> 𝝴
30. char -> LETTER_TK
31. char -> DIGIT_TK
32. char -> UNDERSCORE_TK
33. char -> COLON_TK
34. char -> DASH_TK
```

|                                                                                     | First                                                                                                             | Follow                                                                                                            |
| ----------------------------------------------------------------------------------- | ----------------------------------------------------------------------------------------------------------------- | ----------------------------------------------------------------------------------------------------------------- |
| xmldokument -> element \| xmldecl element                                           | { XML_OPEN_TK, SIMPLE_START_TK }                                                                                  | { $ }                                                                                                             |
| xmldecl -> XML_OPEN_TK XML_VERSION_TK vernumb XML_CLOSE_TK                          | { XML_OPEN_TK }                                                                                                   | { SIMPLE_START_TK }                                                                                               |
| vernumb -> number DOT_TK number                                                     | { DIGIT_TK }                                                                                                      | { XML_CLOSE_TK }                                                                                                  |
| element -> SIMPLE_START_TK name elementshitshit                                     | { SIMPLE_START_TK }                                                                                               | { $, SLASHED_START_TK }                                                                                           |
| elementshitshit -> SLASHED_CLOSE_TK \| SIMPLE_CLOSE_TK elementshit                  | { SLASHED_CLOSE_TK, SIMPLE_CLOSE_TK }                                                                             | { $, SLASHED_START_TK }                                                                                           |
| elementshit -> word endtag \| elements endtag                                       | { SLASHED_START_TK, SIMPLE_START_TK, LETTER_TK, DIGIT_TK, DOT_TK, UNDERSCORE_TK, DASH_TK, COLON_TK ... okrem <> } | { $, SLASHED_START_TK }                                                                                           |
| endtag -> SLASHED_START_TK name SIMPLE_CLOSE_TK                                     | { SLASHED_START_TK }                                                                                              | { $, SLASHED_START_TK }                                                                                           |
| elements -> element \| 𝝴                                                            | { SIMPLE_START_TK, 𝝴 }                                                                                            | { SLASHED_START_TK }                                                                                              |
| name -> LETTER_TK namechars \| UNDERSCORE_TK namechars \| COLON_TK namechars        | { LETTER_TK, UNDERSCORE_TK, COLON_TK }                                                                            | { SLASHED_CLOSE_TK, SIMPLE_CLOSE_TK }                                                                             |
| namechars -> namechar namechars \| 𝝴                                                | { 𝝴, LETTER_TK, DIGIT_TK, DOT_TK, UNDERSCORE_TK, DASH_TK, COLON_TK ... okrem <> }                                 | { SLASHED_CLOSE_TK, SIMPLE_CLOSE_TK }                                                                             |
| namechar -> LETTER_TK \| DIGIT_TK \| DOT_TK \| UNDERSCORE_TK \| DASH_TK \| COLON_TK | { LETTER_TK, DIGIT_TK, DOT_TK, UNDERSCORE_TK, DASH_TK, COLON_TK ... okrem <> }                                    | { SLASHED_CLOSE_TK, SIMPLE_CLOSE_TK, LETTER_TK, DIGIT_TK, DOT_TK, UNDERSCORE_TK, DASH_TK, COLON_TK ... okrem <> } |
| number -> DIGIT_TK numbershit                                                       | { DIGIT_TK }                                                                                                      | { DOT_TK, XML_CLOSE_TK }                                                                                          |
| numbershit -> number \| 𝝴                                                           | { DIGIT_TK, 𝝴 }                                                                                                   | { DOT_TK, XML_CLOSE_TK }                                                                                          |
| word -> char wordshit                                                               | { LETTER_TK, DIGIT_TK, DOT_TK, UNDERSCORE_TK, DASH_TK, COLON_TK ... okrem <> }                                    | { SLASHED_START_TK }                                                                                              |
| wordshit -> word \| 𝝴                                                               | { 𝝴, LETTER_TK, DIGIT_TK, DOT_TK, UNDERSCORE_TK, DASH_TK, COLON_TK ... okrem <> }                                 | { SLASHED_START_TK }                                                                                              |
| char -> LETTER_TK \| DIGIT_TK \| UNDERSCORE_TK \| COLON_TK \| DASH_TK               | { LETTER_TK, DIGIT_TK, DOT_TK, UNDERSCORE_TK, DASH_TK, COLON_TK ... okrem <> }                                    | { SLASHED_START_TK, LETTER_TK, DIGIT_TK, DOT_TK, UNDERSCORE_TK, DASH_TK, COLON_TK ... okrem <> }                  |


### Tabulka prechodov

|                 | XML_OPEN_TK | XML_VERSION_TK | XML_CLOSE_TK | DOT_TK | SIMPLE_START_TK | SIMPLE_CLOSE_TK | SLASHED_START_TK | SLASHED_CLOSE_TK | UNDERSCORE_TK | COLON_TK | DASH_TK | LETTER_TK | DIGIT_TK | others |  $  |
| :-------------- | :---------: | :------------: | :----------: | :----: | :-------------: | :-------------: | :--------------: | :--------------: | :-----------: | :------: | :-----: | :-------: | :------: | :----: | :-: |
| xmldokument     |      2      |                |              |        |        1        |                 |                  |                  |               |          |         |           |          |        |     |
| xmldecl         |      3      |                |              |        |                 |                 |                  |                  |               |          |         |           |          |        |     |
| vernumb         |             |                |              |        |                 |                 |                  |                  |               |          |         |           |    4     |        |     |
| element         |             |                |              |        |        5        |                 |                  |                  |               |          |         |           |          |        |     |
| elementshitshit |             |                |              |        |                 |        7        |                  |        6         |               |          |         |           |          |        |     |
| elementshit     |             |                |              |        |        9        |                 |        9         |                  |       8       |    8     |    8    |     8     |    8     |   8    |     |
| endtag          |             |                |              |        |                 |                 |        10        |                  |               |          |         |           |          |        |     |
| elements        |             |                |              |        |       11        |                 |        12        |                  |               |          |         |           |          |        |     |
| name            |             |                |              |        |                 |                 |                  |                  |      14       |    15    |         |    13     |          |        |     |
| namechars       |             |                |              |   16   |                 |       17        |                  |        17        |      16       |    16    |   16    |    16     |    16    |   16   |     |
| namechar        |             |                |              |   20   |                 |                 |                  |                  |      21       |    23    |   22    |    18     |    19    |        |     |
| number          |             |                |              |        |                 |                 |                  |                  |               |          |         |           |    24    |        |     |
| numbershit      |             |                |      26      |   26   |                 |                 |                  |                  |               |          |         |           |    25    |        |     |
| word            |             |                |              |        |                 |                 |                  |                  |      27       |    27    |   27    |    27     |    27    |   27   |     |
| wordshit        |             |                |              |        |                 |                 |        29        |                  |      28       |    28    |   28    |    28     |    28    |   28   |     |
| char            |             |                |              |        |                 |                 |                  |                  |      32       |    33    |   34    |    30     |    31    |        |     |


## 6. Lexikalny analyzator

Nachadza sa v subore `LexicalAnalyzer.hs`. Analyza sa vykonava pomocou funkcie `analyze`. Vstup je retazec a vystup je zoznam tokenov.

Implementovali sme zotavenie z chyby: ignorujeme retazce na vstupe az kym nenajdeme jeden z ocakavanych retazcov - nachadza sa na 

## 7. Syntakticky analyzator

Nachadza sa v subore `SyntacticAnalyzer.hs` - funkcia `analyze`. Vstup je zoznam tokenov a vystupom je dvojica bolean urcujuci vysledok a zoznam logov. 

Implementovali sme zotavenie z chyby - funkcia `panicModeRecovery`. Vyuzili sme ju na riadkoch 120 a 131. 

CHANGELOG.md
# Revision history for sj-project

## 0.1.0.0 -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


app/LexicalAnalyzer.hs
module LexicalAnalyzer where

import Common
import Data.List
import qualified Data.Map.Strict as Map
import Data.Maybe
import Debug.Trace (trace)

mapping :: Map.Map String Terminal
mapping =
  Map.fromList $
    [ ("<?xml", XmlOpenTk),
      ("version=", XmlVersionTk),
      ("?>", XmlCloseTk),
      (".", DotTk),
      ("<", SimpleStartTk),
      (">", SimpleCloseTk),
      ("/>", SlashedCloseTk),
      ("</", SlashedStartTk),
      ("_", UnderscoreTk),
      (":", ColonTk),
      ("-", DashTk)
    ]
      ++ [([c], LetterTk) | c <- ['a' .. 'z']]
      ++ [([c], LetterTk) | c <- ['A' .. 'Z']]
      ++ [(show d, DigitTk) | d <- [0 .. 9]]

findKeysForInput :: String -> [(String, Terminal)]
findKeysForInput input = filter (\(key, _) -> isJust $ stripPrefix input key) (Map.toList mapping)

inputContainsToken :: String -> Terminal -> Bool
inputContainsToken input token = undefined

analyze :: String -> [Terminal]
analyze input = fst $ foldl go ([], "") input
  where
    go :: ([Terminal], String) -> Char -> ([Terminal], String)
    go (tokens, acc) a =
      let newAcc = acc ++ [a]
       in case findKeysForInput newAcc of
            [] ->
              let accSufix = last newAcc
                  accPrefix = take (length newAcc - 1) newAcc
                  (suffixTokens, suffixSuffix) = go ([], "") accSufix
               in case Map.lookup accPrefix mapping of
                    Just token -> (tokens ++ [token] ++ suffixTokens, suffixSuffix)
                    Nothing -> (tokens, drop 1 newAcc) -- Dropping in case no match in tokens = optimisation
            [(key, token)] -> case stripPrefix key newAcc of
              Nothing -> (tokens, newAcc)
              Just s -> (tokens ++ [token], s)
            _ -> (tokens, newAcc)


app/SyntacticAnalyzer.hs
{-# LANGUAGE LambdaCase #-}

module SyntacticAnalyzer where

import Common
import Data.List
import Data.List.Extra ((!?))
import Data.Maybe (fromJust)

rules :: [(Nonterminal, Maybe [Symbol])]
rules =
  [ (XmlDokument, Just [Right Element]),
    (XmlDokument, Just [Right XmlDecl, Right Element]),
    (XmlDecl, Just [Left XmlOpenTk, Left XmlVersionTk, Right VerNumb, Left XmlCloseTk]),
    (VerNumb, Just [Right Number, Left DotTk, Right Number]),
    (Element, Just [Left SimpleStartTk, Right Name, Right ElementShitShit]),
    (ElementShitShit, Just [Left SlashedCloseTk]),
    (ElementShitShit, Just [Left SimpleCloseTk, Right ElementShit]),
    (ElementShit, Just [Right Word, Right EndTag]),
    (ElementShit, Just [Right Elements, Right EndTag]),
    (EndTag, Just [Left SlashedStartTk, Right Name, Left SimpleCloseTk]),
    (Elements, Just [Right Element]),
    (Elements, Nothing),
    (Name, Just [Left LetterTk, Right NameChars]),
    (Name, Just [Left UnderscoreTk, Right NameChars]),
    (Name, Just [Left ColonTk, Right NameChars]),
    (NameChars, Just [Right NameChar, Right NameChars]),
    (NameChars, Nothing),
    (NameChar, Just [Left LetterTk]),
    (NameChar, Just [Left DigitTk]),
    (NameChar, Just [Left DotTk]),
    (NameChar, Just [Left UnderscoreTk]),
    (NameChar, Just [Left DashTk]),
    (NameChar, Just [Left ColonTk]),
    (Number, Just [Left DigitTk, Right NumberShit]),
    (NumberShit, Just [Right Number]),
    (NumberShit, Nothing),
    (Word, Just [Right Char, Right WordShit]),
    (WordShit, Just [Right Word]),
    (WordShit, Nothing),
    (Char, Just [Left LetterTk]),
    (Char, Just [Left DigitTk]),
    (Char, Just [Left UnderscoreTk]),
    (Char, Just [Left ColonTk]),
    (Char, Just [Left DashTk])
  ]

table :: [(Nonterminal, Terminal, Int)]
table =
  [ (XmlDokument, XmlOpenTk, 2),
    (XmlDokument, SimpleStartTk, 1),
    (XmlDecl, XmlOpenTk, 3),
    (VerNumb, DigitTk, 4),
    (Element, SimpleStartTk, 5),
    (ElementShitShit, SimpleCloseTk, 7),
    (ElementShitShit, SlashedCloseTk, 6),
    (ElementShit, UnderscoreTk, 8),
    (ElementShit, ColonTk, 8),
    (ElementShit, DashTk, 8),
    (ElementShit, LetterTk, 8),
    (ElementShit, DigitTk, 8),
    (ElementShit, SimpleStartTk, 9),
    (ElementShit, SlashedStartTk, 9),
    (EndTag, SlashedStartTk, 10),
    (Elements, SimpleStartTk, 11),
    (Elements, SlashedStartTk, 12),
    (Name, LetterTk, 13),
    (Name, UnderscoreTk, 14),
    (Name, ColonTk, 15),
    (NameChars, DotTk, 16),
    (NameChars, UnderscoreTk, 16),
    (NameChars, ColonTk, 16),
    (NameChars, DashTk, 16),
    (NameChars, LetterTk, 16),
    (NameChars, DigitTk, 16),
    (NameChars, SimpleCloseTk, 17),
    (NameChars, SlashedCloseTk, 17),
    (NameChar, LetterTk, 18),
    (NameChar, DigitTk, 19),
    (NameChar, DotTk, 20),
    (NameChar, UnderscoreTk, 21),
    (NameChar, DashTk, 22),
    (NameChar, ColonTk, 23),
    (Number, DigitTk, 24),
    (NumberShit, DigitTk, 25),
    (NumberShit, DotTk, 26),
    (NumberShit, XmlCloseTk, 26),
    (Word, UnderscoreTk, 27),
    (Word, ColonTk, 27),
    (Word, DashTk, 27),
    (Word, LetterTk, 27),
    (Word, DigitTk, 27),
    (WordShit, UnderscoreTk, 28),
    (WordShit, ColonTk, 28),
    (WordShit, DashTk, 28),
    (WordShit, LetterTk, 28),
    (WordShit, DigitTk, 28),
    (WordShit, SlashedStartTk, 29),
    (Char, LetterTk, 30),
    (Char, DigitTk, 31),
    (Char, UnderscoreTk, 32),
    (Char, ColonTk, 33),
    (Char, DashTk, 34)
  ]

analyze :: [Terminal] -> (Bool, [String])
analyze input = case foldl go ([Right XmlDokument], []) input of
  ([], ss) -> (True, ss)
  (_, ss) -> (False, ss)
  where
    go :: ([Symbol], [String]) -> Terminal -> ([Symbol], [String])
    go (stack, logs) terminal =
      case stack of
        ((Right nonterminal) : xs) ->
          let maybeSymbols =
                find (matchRule nonterminal terminal) table
                  >>= (\(_, _, idx) -> pure idx)
                  >>= (\idx -> rules !? (idx - 1) >>= (\x -> pure (idx, snd x)))
           in case maybeSymbols of
                Nothing -> panicModeRecovery stack logs terminal
                Just (idx, s) ->
                  case s of
                    Just symbols ->
                      let newStack = symbols ++ xs
                       in go (newStack, logs ++ [logMsg ("App rule " <> show idx <> "...") stack terminal]) terminal
                    Nothing ->
                      go (xs, logs ++ [logMsg "Popping...... " stack terminal]) terminal
        ((Left t) : xs) ->
          if t == terminal
            then (xs, logs ++ [logMsg "Popping...... " stack terminal])
            else panicModeRecovery stack logs terminal
        [] -> error "Empty stack before eof"

    panicModeRecovery :: [Symbol] -> [String] -> Terminal -> ([Symbol], [String])
    panicModeRecovery stack logs input = (stack, logs ++ [logMsg "Skip input..." stack input])

    logMsg :: String -> [Symbol] -> Terminal -> String
    logMsg action stack terminal = action <> " \t\tStack: " <> show stack <> ", Input: " <> show terminal

    matchRule :: Nonterminal -> Terminal -> (Nonterminal, Terminal, Int) -> Bool
    matchRule n1 t1 (n2, t2, _) =
      n1 == n2 && t1 == t2
